[ADD] last_purchased_products: sort selection by recent customer history

danal-odoo opened the pull request:
Before, the product selection dropdown did not display the
time of the previous order and products were not sorted based on how
recently they were invoiced for the selected customer. Additionally,
the last invoice time and specific quantity calculations were missing
from the catalog view.

This module enhances the product selection mechanism to assist users in
identifying frequently purchased items. Specific improvements include:
- Displaying the `last_order` time next to the product name in the
  dropdown.
- Sorting products by `last_invoice_date` specific to the selected
  customer (partner).
- Showing `final_qty` and `last_invoice_time` in the product
  catalog kanban view.
- Displaying `invoice_partner_name` next to `last_invoice_time`,
   which is clickable and open last invoice of that property. 

If no customer is selected in the context, the standard selection
behavior and sorting apply.

Technical Details:
- Model `product.product`: Added `last_order` Datetime field,
   `last_invoice_date` Datetime Field, `last_invoice_time` Char field,
   `invoice_partner_name` Char field.
- Model `product.template`: Added `last_order` Datetime field,
   `last_invoice_date` Datetime Field.
- Method `action_open_last_invoice`: To open last invoice of product.
- Method `name_search`: Overridden to order products as `last_invoice_date`.
- Method `_compute_display_name`: Overridden to show `last_order` as to string
   time at right side of product name in selection dropdown.
- Method `_compute_last_order`: to compute `last_order` of shown products in
   selection dropdown in context of selected customer(`partner_id`).
- Method `_compute_last_invoice_date`: to compute `last_invoice_date` to order
   recently invoiced first shown products in selection dropdown in context of
   selected customer(`partner_id`).
- Method `compute_agotime`: to compute given Datetime to string format.

- View `sale_order_views`: Updated `product_template_id` and catalog button
   to pass the selected `partner_id` in the context as `customer`.
- View `account_move_views`: Updated `product_id` and catalog button
   to pass the selected `partner_id` in the context as `customer`.
- View `purchase_order_views`: Updated `product_id` to pass the selected
  `partner_id` in the context as `vendor`.
- View `product_views`: Updated `product_view_kanban_catalog` to display
  `final_qty` (`virtual_available` - `qty_available`) next to `On Hand` and
  `last_invoice_time` after `qty_available` in `product_view_kanban_catalog`.

- Template `order_line`: Updated to show `uomDisplayName` next to `price`.

danal-odoo committed:
[ADD] last_purchased_products: sort selection by recent customer history

Before, the product selection dropdown did not display the
time of the previous order and products were not sorted based on how
recently they were invoiced for the selected customer. Additionally,
the last invoice time and specific quantity calculations were missing
from the catalog view.

This module enhances the product selection mechanism to assist users in
identifying frequently purchased items. Specific improvements include:
- Displaying the `last_order` time next to the product name in the
  dropdown.
- Sorting products by `last_invoice_date` specific to the selected
  customer (partner).
- Showing `final_qty` and `last_invoice_time` in the product
  catalog kanban view.

If no customer is selected in the context, the standard selection
behavior and sorting apply.

Technical Details:
- Model `product.product`: Added `last_order` Datetime field,
   `last_invoice_date` Datetime Field, `last_invoice_time` field.
- Model `product.template`: Added `last_order` Datetime field,
   `last_invoice_date` Datetime Field.
- Method `name_search`: Overridden to order products as `last_invoice_date`.
- Method `_compute_display_name`: Overridden to show `last_order` as to string
   time at right side of product name in selection dropdown.
- Method `_compute_last_order`: to compute `last_order` of shown products in
   selection dropdown in context of selected customer(`partner_id`).
- Method `_compute_last_invoice_date`: to compute `last_invoice_date` to order
   recently invoiced first shown products in selection dropdown in context of
   selected customer(`partner_id`).
- Method `compute_agotime`: to compute given Datetime to string format.

- View `sale_order_views`: Updated `product_template_id` and catalog button
   to pass the selected `partner_id` in the context as `customer`.
- View `account_move_views`: Updated `product_id` and catalog button
   to pass the selected `partner_id` in the context as `customer`.
- View `purchase_order_views`: Updated `product_id` to pass the selected
  `partner_id` in the context as `vendor`.
- View `product_views`: Updated `product_view_kanban_catalog` to display
  `final_qty` (`virtual_available` - `qty_available`) next to `On Hand` and
  `last_invoice_time` after `qty_available` in `product_view_kanban_catalog`.

- Template `order_line`: Updated to show `uomDisplayName` next to `price`.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1083.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1083)

bit-odoo commented on last_purchased_products/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+        "views/product_views.xml"
+    ],
+    "assets": {
+        'web.assets_backend': [
```
Please follow the same quote through the module.

bit-odoo reviewed (commented):
Hello,
I think that your module will not install.
Can you please check if it is working fine or not?

Thanks

danal-odoo committed:
[IMP] last_purchased_products: add test suite for computation and sorting logic

This commit introduces the `TestLastOrderProduct` test suite to ensure the
stability of product sorting and time computation based on sales and
purchase history.

Key scenarios covered:

Last Order Computation:
   - Verifies that confirming a Sale Order updates the `last_order` field.
   - Checks that the 'Ago' time suffix (e.g., '--1h--') is correctly
     appended to `display_name` when the context is active.

Last Invoice Date:
   - Verifies that posting a Customer Invoice updates `last_invoice_date`.
   - Ensures valid date comparisons between fields.

Vendor Bill Sorting:
   - Verifies that posting a Vendor Bill updates `last_invoice_date`
     when in a vendor context.
   - Ensures products from recent bills appear at the top of search results.

Name Search Sorting:
   - Verifies that recently invoiced products appear at the top of the
     list in the `name_search` results.
   - Confirms that sorting falls back to the default (alphabetical)
     when no customer/vendor context is provided.

Time Display Logic:
   - Validates `last_invoice_time` computation.
   - Specifically verifies that invoices created less than a minute ago
     return 'Just Now'.

Context Handling:
   - Ensures logic only triggers when `customer`, `vendor`, or
     `formatted_display_name` keys are present in the context, preventing
     unwanted side effects in standard views.

danal-odoo committed:
[IMP] last_purchased_products: added customer name next to invoice time

Before, there was no customer name next to last invoice time,
it was hard to identify that to which customer this product was
recently invoiced.

this commit add that functionality which will add customer name
to whom then recent invoice  created and that name is clickable
will take to that invoice form.

Technical Details:
- Model `product.product`: Created `invoice_partner_name` Char field
  to store customer name.
- Model `product.product`: Created action `action_open_last_invoice`
  to open last invoice of product.
- View `product_views`: Updated to show customer name next to
  `last_invoice_time`.

bit-odoo reviewed (commented):
Hello,
compute_agotime() most logic is duplicated.
Like this at many places your code get duplicated.

Can you please improve your code?
Thanks
